### ISSUE TITLE
Linq failure on null parameter value

### ISSUE THREAD

**Issue by micmerchant** [nhibernate/nhibernate-core#2833]
Hello! 

First, thank you for fixing our issue https://github.com/nhibernate/nhibernate-core/issues/2805 in your develop branch. But as it was difficult to identify further regression bugs, due to the major problems we had with the solved issue, we decided to give the develop branch a try. 

I'm not sure if it is okay to create an issue for an unreleased version ... simply because it is work in progress .. but there is one that is causing troubles in our application (at least in our unit tests). But I guess, the sooner you know about potential problems, the better.

So, I've added an unit test, which fails with the following exception in the dev branch, but not in the current stable release (5.3.8):
```
NHibernate.Test.Linq.LinqQuerySamples.ReplaceFunctionWithNullArgument

System.ArgumentNullException : A type specific Set(name, val) should be called because the Type can not be guessed from a null value.
Parameter name: val
   at NHibernate.Impl.AbstractQueryImpl.SetParameter(String name, Object val) in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate\Impl\AbstractQueryImpl.cs:line 307
   at NHibernate.Linq.DefaultQueryProvider.SetParameters(IQuery query, IDictionary`2 parameters) in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate\Linq\DefaultQueryProvider.cs:line 268
   at NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query) in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate\Linq\DefaultQueryProvider.cs:line 215
   at NHibernate.Linq.DefaultQueryProvider.Execute(Expression expression) in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate\Linq\DefaultQueryProvider.cs:line 94
   at NHibernate.Linq.DefaultQueryProvider.Execute[TResult](Expression expression) in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate\Linq\DefaultQueryProvider.cs:line 101
   at System.Linq.Queryable.First[TSource](IQueryable`1 source)
   at NHibernate.Test.Linq.LinqQuerySamples.ReplaceFunctionWithNullArgument() in C:\git\Repositories\Tools\NHibernate\Default\Core\src\NHibernate.Test\Linq\LinqQuerySamples.cs:line 1967



-----

One or more child tests had errors
  Exception doesn't have a stacktrace
```
`LinqQuerySamples`:
```
[Test]
public void ReplaceFunctionWithNullArgument()
{
	var query = from e in db.Employees
				select e.FirstName.Replace(e.LastName, null);
	var result = query.First();
	Assert.That(result, Is.Null, "Expected REPLACE(FirstName, LastName, NULL) to be null");
}
```

Thanks in advance,
best regards
Michael

**Comment by bahusoid** [comment]
Did I get it right - this test works with 5.3.8 release and fails with 5.3.9-dev version (5.3.x branch)?

**Comment by micmerchant** [comment]
Exactly.

**Comment by fredericDelaporte** [comment]
Regression from #2793 (checked by branching before and after, running the test in both branches).

About the test, expecting the result to be `null` is a SQL Server behavior expectation (and maybe of some other database engines). From a Linq-To-Object standpoint, we should instead expect a non-null string. So I would rather avoid testing this behavior which might be database vendor specific.
I would rather write the test as:
```c#
[Test]
public void ReplaceFunctionWithNullArgument()
{
	var query = from e in db.Employees
	            select e.FirstName.Replace(e.LastName, null);
	List<string> results = null;
	Assert.That(
		() =>
		{
			results = query.ToList();
		}, Throws.Nothing, "Expected REPLACE(FirstName, LastName, NULL) to be supported");
	Assert.That(results, Is.Not.Null);
}
```

**Comment by fredericDelaporte** [comment]
Relabeling as a "fix" type since this issue has not hit a released version: there is no need to surface it in the release notes.

**Comment by fredericDelaporte** [comment]
Fixed by #2836
